Add documentation for dashboards assistant plugin (#6370)
opensearch-project/documentation-website: _dashboards-assistant/index.md

---
layout: default
title: Dashboards Assistant
nav_order: 1
has_children: false
has_toc: false
nav_exclude: true
permalink: /dashboards-assistant/
redirect_from:
  - /dashboards-assistant/index/
---

This is an experimental feature and is not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, please checkout [GitHub repo](https://github.com/opensearch-project/dashboards-assistant).
{: .warning}

# Dashboards assistant

> The OpenSearch Assistant Dashboards plugin lets you have an OpenSearch assistant to help dashboards users use OpenSearch. With dashboards assistant, user can quickly gain insights on their operational data just with nature language without the need to specialized in query tools and skills.

Reviewer: Note: I'll add the missing link.

> The OpenSearch Assistant Dashboards plugin lets you have an OpenSearch assistant to help dashboards users use OpenSearch. With dashboards assistant, user can quickly gain insights on their operational data just with nature language without the need to specialized in query tools and skills.

Reviewer 2: Do we want to talk about the plugin in particular here? The plugin piece is part of the OpenSearch Assistant toolkit to create an chat interface for a LLM-powered OpenSearch Assistant to help users interact with OpenSearch.

Under the response, the second from the left is regenerate button, click it will regenerate answer for the same question. The previous answer will be replaced with the new answer both on the UI and chat history.

### The suggestions by the assistant
Under the response, we also provide suggestions that we think user may be interested to ask. User can click one of them to see the response.

<img width="600px" src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/suggestions.png" alt="The assistant suggestions">

### Rate the response
Each response contains a thumb up and thumb down icon button, you can click on the button to feedback whether you are satisfied with the response.

<img width="600px" src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/rate.png" alt="Rate the response">

### See how the response was generated
Under the available suggestions, there will be a link “How was this generated?”, click this link will show which tools are triggered to generate this response. It can be used for trouble shooting. There could be multiple steps if multiple tools are involved. Each step will display tool name with input and output of the tool.

> <img width="400px" src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/traces.png" alt="How was the response generated">

Reviewer: ```suggestion
<img width="400px" src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/traces.png" alt="Snapshot of response generation details">
```

> ### Continue with a past conversation

Reviewer: ```suggestion
### Resuming previous conversations

```

> #### View past conversation

Reviewer: ```suggestion

```

> Click the clock icon to open the conversation history panel. The clock icon will become active status and conversations will be displayed clicked after clicked.

Reviewer: ```suggestion
To view a previous conversation, select the clock icon to open the conversation history panel and display the chat history. A snapshot of a conversation history is shown in the following image.
```

> <img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/conversation-history-entry.png" alt="Show conversation history">

Reviewer: ```suggestion
<img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/conversation-history-entry.png" alt="Snapshot of conversation history">
```

> Search conversations by title:

Reviewer: ```suggestion
The conversation history can be searched by title, as shown in the following snapshot. 
```

<img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/conversation-history-list.png" alt="List of conversations">

> #### Manage past conversation

Reviewer: ```suggestion
#### Editing and deleting previous conversations

```

> Click the pencil icon to edit conversation name, type the new conversation name in the below input, then click “Confirm name” button. The new conversation name will be saved.

Reviewer: ```suggestion
Select the pencil icon to edit a conversation name and rename it. Select the **Confirm name** button to save the new name. A snapshot of the edit interface is shown in the following image.
```

> <img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/edit-conversation-title.png" alt="Edit the conversation title">

Reviewer: ```suggestion
<img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/edit-conversation-title.png" alt="Editing a conversation name">
```

> Click the trash icon to delete conversation, the delete confirmation modal will be displayed after clicked. Click the “Delete conversation” button to confirm the conversation delete operation. The conversation will be deleted after “Delete conversation” button clicked.

Reviewer: ```suggestion
Select the trash icon to delete conversation. The confirmation dialog will appear. Select “Delete conversation”. The conversation is now deleted from your chat history. A snapshot of the delete interface is shown in the following image.
```

> <img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/delete-conversation.png" alt="Delete the conversation">

Reviewer: ```suggestion
<img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/delete-conversation.png" alt="Deleting a conversation">
```

> ### Share a conversation via notebook

Reviewer: ```suggestion
### Sharing a conversation through Notebooks

```

> Click `Save to notebook` button to save your conversation to notebook. All messages between you and LLM model will be saved to observability notebook after entering name for notebook and confirm.

Reviewer: ```suggestion
You can use [Notebooks]({{site.url}}{{site.baseurl}}/observing-your-data/notebooks/) to save your conversations. To use this option, select `Save to notebook` from the dropdown menu on the right-side of **OpenSearch Assistant**. Enter a name for the notebook and save it. All conversations (questions and responses) between you and the large language model (LLM) will be saved to this notebook. The following snapshot shows a save notebook interface. 
```

> <img width="400px" src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/save-conversation-to-notebook.png" alt="Save a conversation to notebook">

Reviewer: ```suggestion
<img width="400px" src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/save-conversation-to-notebook.png" alt="Saving a conversation through Notebooks">
```

> Now you can find your conversation in observability notebook lists and others can see it as well.

Reviewer: ```suggestion
You can open your saved notebook or view a list of other notebooks by going to **Observability** > **Notebooks**. The following snapshot shows the the Notebooks interface in OpenSearch Dashboards.
```

> <img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/conversation-in-notebook.png" alt="Conversation saved to notebook">

Reviewer: ```suggestion
<img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/conversation-in-notebook.png" alt="List of conversations saved in Notebooks">
```

Reviewer: Add new section 

## Relate information

The following resources provide more information about the OpenSearch Assistant Toolkit and related components:

- [OpenSearch adds a new generative AI assistant toolkit](https://opensearch.org/blog/opensearch-adds-new-generative-ai-assistant-toolkit/)
- <other resources>

{: .warning}

For more information about ways to enable experimental features, see [Experimental feature flags]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/experimental/).
{: .note}
{: .warning}

# OpenSearch Assistant for OpenSearch Dashboards
Introduced 2.12
{: .label .label-purple }

The [OpenSearch Assistant toolkit](<insert-ML-cross-ref-link>) helps you create AI-powered assistants for OpenSearch Dashboards without requiring specialized query tools or skills.

## Enabling OpenSearch Assistant

To enable **OpenSearch Assistant** in OpenSearch Dashboards, locate your copy of the `opensearch_dashboards.yml` file and set the following option:
```
assistant.chat.enabled: true
```
{% include copy-curl.html %}
Then configure the root `agent_id` through the following API:
```
PUT .plugins-ml-config/_doc/os_chat
{
    "type":"os_chat_root_agent",
    "configuration":{
        "agent_id": "your root agent id"
    }
}

> ```

Reviewer 2: We should call out this is a system index, so in security enabled domains, the user must be super admin to do this.
They can use this guide to make super admin calls: https://opensearch.org/docs/latest/security/configuration/system-indices/

Reviewer: @lezzago Please review the following callout text to make sure I've captured your comment accurately:

_This example is a system index. In security-enabled domains, only super administrators have access to execute this code. For details about making super admin calls, see the [System indexes](https://opensearch.org/docs/latest/security/configuration/system-indices/) guide. For access permission, contact your IT administrator._

{% include copy-curl.html %}
Next, restart the OpenSearch Dashboards server. Following a successful restart, **OpenSearch Assistant** should appear in the top-right corner of the OpenSearch Dashboards interface. An example view is shown in the following image.

<img width="450px" src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/entry.png" alt="OpenSearch Assistant icon">

## Configuring the plugin
Go to the [Flow Framework plugin](https://github.com/opensearch-project/flow-framework) documentation to get the default chatbot template. You can modify this template to use your own model and customize the chatbot tools. Go to the [Getting started guide](https://github.com/opensearch-project/dashboards-assistant/blob/main/GETTING_STARTED_GUIDE.md) for step-by-step instructions. 

## Using OpenSearch Assistant in OpenSearch Dashboards
The following sections guide you through using OpenSearch Assistant in OpenSearch Dashboards.

### Start the conversation

> Start the conversation by entering a question in the **Ask a question** search box or using the shortcut `ctrl + /`. An interface snapshot is shown in the following image.

Reviewer: Yes, I'll make changes throughout.

<img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/conversation-history-entry.png" alt="Snapshot of conversation history">

The conversation history can be searched by title, as shown in the following snapshot. 

<img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/conversation-history-list.png" alt="List of conversations">

#### Editing and deleting previous conversations

Select the pencil icon to edit a conversation name and rename it. Select the **Confirm name** button to save the new name. A snapshot of the edit interface is shown in the following image.

<img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/edit-conversation-title.png" alt="Editing a conversation name">

Select the trash icon to delete conversation. The confirmation dialog will appear. Select “Delete conversation”. The conversation is now deleted from your chat history. A snapshot of the delete interface is shown in the following image.

<img src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/delete-conversation.png" alt="Deleting a conversation">

### Sharing a conversation through Notebooks

> You can use [Notebooks]({{site.url}}{{site.baseurl}}/observing-your-data/notebooks/) to save your conversations. To use this option, select `Save to notebook` from the dropdown menu on the right-side of **OpenSearch Assistant**. Enter a name for the notebook and save it. All conversations (questions and responses) between you and the large language model (LLM) will be saved to this notebook. The following snapshot shows a save notebook interface. 

Reviewer 3: ```suggestion
You can use [Notebooks]({{site.url}}{{site.baseurl}}/observing-your-data/notebooks/) to save your conversations. To use this option, select `Save to notebook` from the dropdown menu to the right of **OpenSearch Assistant**. Enter a name for the notebook and save it. All conversations (questions and responses) between you and the large language model (LLM) will be saved to this notebook. The following snapshot shows a save notebook interface. 
```

> You can use [Notebooks]({{site.url}}{{site.baseurl}}/observing-your-data/notebooks/) to save your conversations. To use this option, select `Save to notebook` from the dropdown menu on the right-side of **OpenSearch Assistant**. Enter a name for the notebook and save it. All conversations (questions and responses) between you and the large language model (LLM) will be saved to this notebook. The following snapshot shows a save notebook interface. 

Reviewer 3: "screenshot" instead of "snapshot"?

<img width="400px" src="{{site.url}}{{site.baseurl}}/images/dashboards-assistant/save-conversation-to-notebook.png" alt="Saving a conversation through Notebooks">

> You can open your saved notebook or view a list of other notebooks by going to **Observability** > **Notebooks**. The following snapshot shows the the Notebooks interface in OpenSearch Dashboards.

Reviewer 3: ```suggestion
You can open your saved notebook or view a list of other notebooks by going to **Observability** > **Notebooks**. The following snapshot shows the Notebooks interface in OpenSearch Dashboards.
```

> You can open your saved notebook or view a list of other notebooks by going to **Observability** > **Notebooks**. The following snapshot shows the the Notebooks interface in OpenSearch Dashboards.

Reviewer 3: "screenshot" instead of "snapshot"?
